Pull review/1414

hvacengi opened the pull request:
PR with progress so far on for my review.

tomekpiotrowski committed:
Ranges

TDW89 committed:
removed outdated info

Comrange has been depreciated
I don't know when / if ship:vesselname worked but it doesn't exist now / throws an unknown suffix error now.

TDW89 committed:
fixed typo

fixed typo (second y attribute changed for z)

tomekpiotrowski committed:
Fix newlines in ArchiveFile

TDW89 committed:
fixed typo

Changed `RESOURCESLEX` structure (in table) to lexicon from list

TDW89 committed:
removed outdated info

changed the wording in Archive p2 from archive stored in plugin directory to stored in ship/scripts. removed p7 & 8 as they were outdated references to comrange.

TDW89 committed:
added info about lex() alias 

added info about lex() alias

hvacengi committed:
Merge branch 'develop' of https://github.com/KSP-KOS/KOS into pull-review/1323

Conflicts:
	doc/source/structures/collections/list.rst

hvacengi committed:
Minor refactor

EnumerableValue.cs
* Refactor Enum to be InnerEnumerable
* Code formatting changes
QueueValue.cs, RangeValue.cs, StackValue.cs
* Refactor InnerEnum to InnerEnumerable

hvacengi committed:
Rework constructor function

Suffixed.cs
* Rework the logic in the range constructor.
* Now uses the GetInt helper function so it throws argument mismatch
exceptions
* Removed the two arrays in favor of named parameters.
* Changed argument count exception to be KOSArgumentMismatchException
RangeValue.cs
* Add static references to default constructor parameters.

tomekpiotrowski committed:
Merge pull request #1460 from TDW89/develop

minor doc fixes

hvacengi committed:
Documentation changes

range.rst
* Clarify the whole number behavior of range types.
list.rst
* Fix the reference to clear (which was previously reported as an
attribute that returned the empty list)

hvacengi committed:
Merge branch 'pull-review/1323' into develop

hvacengi committed:
Merge branch 'develop' of https://github.com/KSP-KOS/KOS into pull-review/1414

Conflicts:
	src/kOS.Safe/Encapsulation/EnumerableValue.cs
	src/kOS.Safe/Encapsulation/ListValue.cs

hvacengi committed:
Update nomenclature workflow

Add Nomenclature names to new classes and add a "subclass" names to
some
types with which inherit from other
types.

DMScienceExperimentFields.cs
* Change namespace to
kOS.Addons.OrbitalScience to match folder path

hvacengi committed:
Update nomenclature workflow

KOSNomenclature.cs
* Remove exception throwing from inside try/catch in PopulateMapping.
Bootstrapper.cs
* Added call to PopulateMapping from within Bootstrapper to get the
mappings when the game launches.
CPU.cs
* Shutdown the processor if there are nag messages, breaking all
execution.
kOSProcessor.cs
* Remove reference to PopulateMapping

Dunbaratu commented on doc/source/structures/collections/list.rst:
```diff
@@ -41,6 +41,9 @@ Structure
 
+        * - All suffixes of :struct:`Enumerable`
+          -
+          - :struct:`List` objects are a type of :struct:`Enumerable`
```
Man I should really look into whether or not Sphinx has a sort of include operator that can be used for cases like this.  We have a lot of them.

Dunbaratu replied:
Technically, it's a kind of `Collection` now, and a `Collection`, in turn, is a kind of `Enumerable`.  Thus the 'CLEAR' suffix is part of `Collection`.  But I'm not sure we want this to jump through two levels of link before we get to LENGTH, CONTAINS, And so on.

(the same issue arises with the other collections - queue, etc).

tomekpiotrowski replied:
I wanted to avoid an extra page just for a single method.

Dunbaratu replied:
Okay I can see that.  It's just that when people do `PRINT LIST(1,2,3):INHERITANCE.` they'll see the Collection type in the printout that's not documented, between List and Enumerable.  Maybe that's not such a big deal?

Dunbaratu commented on doc/source/structures/collections/range.rst:
```diff
@@ -0,0 +1,91 @@
+
+- :code:`RANGE(FROM, TO, STEP)`
+  Will create a sequence of numbers that starts with `FROM` and contains all numbers `N + STEP` such that `N` belongs to the sequence and that are smaller than `TO`. For example
+  :code:`RANGE(3,8,2)` will contain numbers 3, 5 and 7. If `FROM` > `TO` then the sequence will be descending. For example :code:`RANGE(2,-9,3)` will contain numbers 2, -1, -4 and -7.
```
I'm not sure how I feel about this.  I didn't look at tomek's range feature in detail until now and I'm not sure I like it not including the endpoint and yet using the terminology "from...to".

"Count from 1 to 10" doesn't mean to stop at 9.  Even when a mathematical concept uses [0..n-1] rather than [1..n], the way you usually speak that aloud is "zero to N minus one", not "zero to N".

I get the mathematical reason to not include the endpoint, but maybe the terminology should be a bit different somehow.  Calling it "TO" really gives the wrong message.  You're not counting TO this value, but counting to just shy of it.  Maybe call them "begin" and "end"?

tomekpiotrowski replied:
Feel free to change the naming. Python uses start/stop for example.

hvacengi replied:
I was actually going to discuss the same thing, until I started looking and found that it's the same limitation of python.  And because the default is to start at 0 this method allows you to work with zero based indices:

```
set l to list("one", "two", "three").
for i in range(l:length) {
    print l[i] + " (" + i + ")".
}
```

I'm not sure that changing the closing parameter to "end" will really convey that it is an "exclusive" limit any better than "to".  Like so many things, it's difficult to convey the mathematical meaning in a single word.  I suppose for the purpose of documentation the name of the parameter is only useful for helping the KS users understand the meaning, so we could simple change it to "from_inclusive" and "to_exclusive" (though that's implying an affinity towards "_" deliminated naming)

Dunbaratu replied:
I think this is a case where ambiguity would actually be an improvement.  "Start at 1, Stop at 10" could mean either [1..9] or [1..10] depending on interpretation, or "Begin at 1, End at 10" could mean either [1..9] or [1..10] depending on interpretation, but "From 1 to 10" only means [1..10].  It's an unambiguous meaning, but unfortunately its unambiguously the wrong meaning.  An ambiguous meaning that has to be specified in the docs would be an improvement over an unambiguous one that has to be contradicted in the docs.

Dunbaratu commented on doc/source/structures/collections/range.rst:
```diff
@@ -0,0 +1,91 @@
+There are 3 ways of constructing a :struct:`Range`:
+
+- :code:`RANGE(FROM, TO, STEP)`
+  Will create a sequence of numbers that starts with `FROM` and contains all numbers `N + STEP` such that `N` belongs to the sequence and that are smaller than `TO`. For example
```
The phrase "that are smaller than" is wrong when counting backward with a negative STEP.  Then the endpoint is at the low end of the range, not the high end.

tomekpiotrowski replied:
I thought that things will be clear by looking at the example below, but sure, we can add an extra sentence explaining that case.

hvacengi replied:
Oh, I forgot I was planning on making a revision to that line in my original review.  I was going to rephrase it to "belongs to the sequence and is between `FROM` and `TO` (including `FROM`)"

Dunbaratu commented on doc/source/structures/collections/list.rst:
```diff
@@ -50,33 +53,18 @@ Structure
           - None
           - remove item by index
-        * - :attr:`CLEAR`
+        * - :meth:`CLEAR()`
```
Technically this clear suffix is part of `Collection` now.  But I don't know if we want to have that much extra clicking to get to the `Enumerable` suffixes (see comment above).

(the same issue arises with the other collections - queue, etc).

hvacengi replied:
Lists are the only thing that inherit from CollectionValue actually (currently).  Queue and Stack inherit directly from EnumerableValue.  So List is really the only structure where `Clear` shows up.

Dunbaratu replied:
What's the rationale for not being able to `CLEAR` a queue or a stack?

Dunbaratu commented on src/kOS.Safe/Encapsulation/EnumerableValue.cs:
```diff
@@ -30,23 +34,44 @@ IEnumerator IEnumerable.GetEnumerator()
+            AddSuffix("LENGTH",          new NoArgsSuffix<ScalarValue>(() => InnerEnumerable.Count()));
+            AddSuffix("CONTAINS",        new OneArgsSuffix<BooleanValue, T>((n) => Contains(n)));
+            AddSuffix("EMPTY",           new NoArgsSuffix<BooleanValue>(() => !InnerEnumerable.Any()));
+            AddSuffix("DUMP",            new NoArgsSuffix<StringValue>(() => new StringValue(ToString())));
```
I prefer the consistency in the deleted formatting here.  Either you prefer having columns over having tight spacing, or you prefer having tight spacing over having columns, but this is a weird half and half mixture.  You removed the spaces that made column 3 line up, but kept the ones that made column 2 line up.

tomekpiotrowski replied:
Yeah, normally I never do those kind of indents, but I'm trying to maintain them when seeing one in the code. Apparently I'm not very good at this ;P

hvacengi replied:
Actually this was my change.  I was attempting to compromise.  I do not like the extra white space  that comes with the indentation, but others think it helps to clarify.  So I thought that it would end up being OK if the methods themselves weren't indented, but the constructor was.  I'd rather just eliminate the indentation outright.

Dunbaratu replied:
Whatever your opinion on the clarifying (ahem, not "extra") spaces are, the original way is consistent with how the other InitializeSuffixes() throughout the mod are doing it.  I get my nice formatting overridden in this mod so much that the one time there actually does seem to be a practice in place that goes the other way and does it the way I like, I'm going to stick up for it.

Dunbaratu commented on src/kOS.Safe/Encapsulation/EnumerableValue.cs:
```diff
@@ -1,26 +1,30 @@
+        protected EnumerableValue(string label, TE enumerable)
         {
-            Collection = collection;
+            this.label = label;
```
hmm.  What might be better is to pass in the GetType() of the derived type rather than a string label.  Then set the string label using KOSNomenclature.GetKOSName().  Thus we guarantee we are printing a message to the user using the type names that the rest of kOS is showing them for the type, and we future-proof it against future decisions to change that kOS type name.

tomekpiotrowski replied:
:+1:

hvacengi replied:
Yeah, I couldn't make those types of changes in the initial review of Tom's PRs because I didn't have Nomenclature.  I intend to go through and find this kind of thing if possible.  I also want to make Argument Mismatch exceptions show the new Nomenclature names.

Dunbaratu replied:
Excellent idea about the argument mismatch exceptions.

Dunbaratu commented on src/kOS.Safe/Encapsulation/RangeValue.cs:
```diff
@@ -0,0 +1,117 @@
+    public class RangeValue : EnumerableValue<ScalarIntValue, Range>
+    {
+        private const string DumpFrom = "from";
+        private const string DumpTo = "to";
```
See comment above in the .rst doc about it being strange to use the words "from" and "to" to describe a range with an open endpoint that doesn't actually go all the way TO the "to" value.

Dunbaratu commented on src/kOS.Safe/Encapsulation/QueueValue.cs:
```diff
@@ -13,71 +13,59 @@ public QueueValue() : this(new Queue<T>())
             var result = new DumpWithHeader
             {
-                Header = "QUEUE of " + Collection.Count() + " items:"
+                Header = "QUEUE of " + InnerEnumerable.Count() + " items:"
```
Hmm. Maybe get the name from KOSNomenclature.GetKOSName(GetType()) instead of hardcoding the word "QUEUE" here, thus putting all the configuration of what we call this thing when talking to the user into one single place - the attribute.  That same thing would apply to all the other places we hardcode the type names like this.

Come to think of it, we could add that as a get property to the base class `Structure`, so all Structure's can just use KOSName as a property.

hvacengi replied:
Adding that method to Structure is a great idea.  If we make the change here, we'll want to make sure that it's handled throughout the serialization methods.

Dunbaratu commented on src/kOS/Module/Bootstrapper.cs:
```diff
@@ -24,6 +24,7 @@ public void Start()
 
             CheckForLegacyArchive();
 
+            KOSNomenclature.PopulateMapping(typeof(kOS.Safe.Encapsulation.Structure).Assembly, this.GetType().Assembly);
```
Doing it from here is definitely an improvement.

Dunbaratu commented on doc/source/structures/collections/lexicon.rst:
```diff
@@ -46,7 +46,7 @@ Constructing a lexicon
 
 If you wish to make your own lexicon from scratch you can do so with the
-LEXICON() built-in function::
+LEXICON() or LEX() built-in function::
```
I think we should add a "LEX" alias KOSNomenclature for Lexicon.cs to go with this.  See GeoCoordinates.cs as an example of the way that can be done.

hvacengi replied:
Good catch.

Dunbaratu commented on src/kOS.Safe/Encapsulation/RangeValue.cs:
```diff
@@ -0,0 +1,117 @@
+        private void InitializeRangeSuffixes()
+        {
+            AddSuffix("FROM", new NoArgsSuffix<ScalarValue>(() => InnerEnumerable.From));
+            AddSuffix("TO", new NoArgsSuffix<ScalarValue>(() => InnerEnumerable.To));
```
Ah crud - From/To the terminology isn't just in the docs.  It's also exposed here.  That means it bugs me a bit more than it otherwise would.  Docs can be changed or argued about post-release more easily than what the suffix is actually called can be.  Once we release with the exposed name in the suffix then we have to keep it that way or break scripts.

hvacengi committed:
KOSCastException using nomenclature

KOSCastException.cs
* use the KOSNomenclature methods to get the "type" name for the
mismatch exception
CalculatorString.cs
* CastException references StringValue now
Opcode.cs
* Logic not now references BooleanValue
StringValue.cs
* Because index accepts a double, not just an int, now references
ScalarValue for cast exception
FunctionBase.cs
* Now references ScalarValue
